extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1715,12 +1725,95 @@ private void publishSegmentsIfNeeded(
   {
     if (queryKernel.isSuccess() && MSQControllerTask.isIngestion(task.getQuerySpec())) {
       final StageId finalStageId = queryKernel.getStageId(queryDef.getFinalStageDefinition().getStageNumber());
+      queryDef.getFinalStageDefinition().getClusterBy();
 
       //noinspection unchecked
       @SuppressWarnings("unchecked")
       final Set<DataSegment> segments = (Set<DataSegment>) queryKernel.getResultObjectForStage(finalStageId);
+      DataSchema dataSchema = ((SegmentGeneratorFrameProcessorFactory) queryKernel.getStageDefinition(finalStageId)
+                                                                                  .getProcessorFactory()).getDataSchema();
+
+      ShardSpec shardSpec = segments.isEmpty()
+                            ? null
+                            : segments.stream()
+                                                                 .findFirst()
+                                                                 .get()
+                                                                 .getShardSpec();
+      List<String> partitionDimensions = Collections.emptyList();
+
+      if (shardSpec != null && (Objects.equals(shardSpec.getType(), ShardSpec.Type.SINGLE)
+                                || Objects.equals(shardSpec.getType(), ShardSpec.Type.RANGE))) {
+        partitionDimensions = ((DimensionRangeShardSpec) shardSpec).getDimensions();
+      }
+
+      Function<Set<DataSegment>, Set<DataSegment>> compactionStateAnnotateFunction = compactionStateAnnotateFunction(
+          task(),
+          context.jsonMapper(),
+          dataSchema,
+          partitionDimensions
+      );
       log.info("Query [%s] publishing %d segments.", queryDef.getQueryId(), segments.size());
-      publishAllSegments(segments);
+      publishAllSegments(compactionStateAnnotateFunction.apply(segments));
+    }
+  }
+
+  public static Function<Set<DataSegment>, Set<DataSegment>> compactionStateAnnotateFunction(
+      MSQControllerTask task,
+      ObjectMapper jsonMapper,
+      DataSchema dataSchema,
+      List<String> partitionDimensions
+  )
+  {
+    final boolean storeCompactionState = task.getContextValue(
+        Tasks.STORE_COMPACTION_STATE_KEY,
+        Tasks.DEFAULT_STORE_COMPACTION_STATE
+    );
+
+    if (storeCompactionState) {
+      IndexSpec indexSpec = task.getQuerySpec().getTuningConfig().getIndexSpec();
+      GranularitySpec granularitySpec = dataSchema.getGranularitySpec();
+      DimensionsSpec dimensionsSpec = dataSchema.getDimensionsSpec();
+      Map<String, Object> transformSpec = dataSchema.getTransformSpec() == null
```

Comment:
Curious to know why this is required for compaction state.

Reply:
All these fields are captured when setting compaction state in the [native flow](https://github.com/apache/druid/blob/54d0e482dc560f42194c613d1e1bcd5310f281c2/indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java#L586).